### ISSUE TITLE
segfault in _Acquisition.so's `Wrapper_findattr`; object lifetime issue?

### ISSUE THREAD

**Issue by jamadden** [zopefoundation/Acquisition#5]
Recently on 64-bit Linux with Python 2.7.8 we've been seeing a web application crash the python interpreter with segmentation faults like this:

```
segfault at 7f8619c3e054 ip 00007f864a5debb5 sp 00007fff347f37b0 error 4 in _Acquisition.so[7f864a5da000+8000]
```

With the base address of 0x7f864a5da000 and the instruction pointer of 0x7f864a5debb5, that gives us our crashing offset into the binary _Acquisiton.so of 0x4bb5. gdb and addr2line identify that as line 487 in _Acquisition.c:

```C
  if ((*name=='a' && name[1]=='q' && name[2]=='_') ||
```

Looking at the disassembly backs that up:

```assembly
   0x0000000000004ba6 <+86>:	cmovne %rdx,%rbx
   0x0000000000004baa <+90>:	test   $0x10000000,%eax
   0x0000000000004baf <+95>:	jne    0x52c0 <Wrapper_findattr+1904>
   0x0000000000004bb5 <+101>:	movzbl (%rbx),%eax
   0x0000000000004bb8 <+104>:	cmp    $0x61,%al
   0x0000000000004bba <+106>:	je     0x4de0 <Wrapper_findattr+656>
```

In other words, the crash is on dereferencing `name` itself (because the next instruction is to compare it with 'a'). (FWIW, this is in Acquisition 4.2, but I compared the disassembly and compilation with a build of Acquisition 4.0.3 from November and they are the same.)

Where did name come from? Here's the lines leading up to 487, the very first lines in the function:

```C
  PyObject *r, *v, *tb, *tmp;
  char *name="";

  if (PyString_Check(oname)) name=PyString_AS_STRING(oname);
  if (PyUnicode_Check(oname)) {
    tmp=PyUnicode_AsASCIIString(oname);
    if (tmp==NULL) return NULL;
    name=PyString_AS_STRING(tmp);
    Py_DECREF(tmp);
  }
  if ((*name=='a' && name[1]=='q' && name[2]=='_') ||
```

Now, from the error message, we know that `name` isn't a null pointer. We also know that it's not the constant value that `name` is initialized to (because it's out of the address space that the .so occupies).

If `oname` had been a `str` object, `name` would directly be pointing to the `oname->ob_sval` buffer, because that's what `PyString_AS_STRING` does; As the documentation says, "The pointer refers to the internal buffer of `string`, not a copy." Presumably that buffer isn't invalid (could it be?), and there's no way for that to change during this function.

On the other hand, if `oname` was a unicode object, a temporary string is allocated with `PyUnicode_AsASCIIString`. This returns a new reference, and the code then decrements the reference, presumably releasing the string. But at that point, `name` is a pointer to the `ob_sval` of a now-released string.

The only hypothesis I have to explain the crash is this dangling pointer.  Depending on the internal state of the Python (arena) memory allocator, the page that the unicode string was on may have been `munmap` or simply `free`d back to the operating system. Most of the time it works fine because there's nothing that would cause that memory to get reused (corrupting the string) or unmapped, but sometimes, under certain conditions (memory pressure?), the dangling pointer crashes.

Does this sound plausible? If so, is the fix simply to copy the string before dereferencing the temporary string? I can submit a PR for that.

**Comment by jamadden** [comment]
By building a new Python interpreter with `PYMALLOC_DEBUG` set, I was able to see that the `name` string does become a dangling pointer after the `Py_DECREF`. In this build, the memory of deleted objects is specifically scribbled over at delete time:

```C
  if (PyUnicode_Check(oname)) {
    tmp=PyUnicode_AsASCIIString(oname);
    if (tmp==NULL) return NULL;
    name=PyString_AS_STRING(tmp);
	printf("\n$$$ Got unicode; name %s\n", name);
    Py_DECREF(tmp);
	printf("\n$$$ DECREF uni ; name %s\n", name);
  }
```

```python
>>> import Acquisition._Acquisition
[134190 refs]
>>> class X(object): pass
...
[134213 refs]
>>> x = X()
[134216 refs]
>>> w = Acquisition.ImplicitAcquisitionWrapper(x, None)
[134219 refs]
>>> Acquisition.aq_get(w, u'foo')

$$$ Got unicode; name foo

$$$ DECREF uni ; name ��������������������
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: foo
[134248 refs]
```

So it seems like basically luck is what's making the unicode case work now. I can't prove it's the also the cause of the segfault because I haven't trapped it under a debugger and it only happens pretty rarely in production---but we do run with `from __future__ import unicode_literals` so most calls to `getattr` would be passing unicode strings.

I'll work on a PR to fix this. Unfortunately, I'm not sure how to test for it; any suggestions would be welcome. Alas, it's going to be a bit bigger of a change than just copying the `name` string, because of needing to free it again, and there are so many `return` points in that function. It looks to me like a wrapper function will be the way to go, but I'm open to any other ideas.

**Comment by tseaver** [comment]
I believe your analysis is correct, ant the fix you propose in #6 looks correct.
